### PR TITLE
MM-49808: Starting call from profile popover will start call in DM

### PR DESCRIPTION
**Pull request by sinansonmez**
#### Summary
This PR introduces a change to start call in DM when users starting call from profile popover

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/22139
Fixes https://mattermost.atlassian.net/browse/MM-49808

#### Release Note
```release-note
Start a call in DM when users starting call from profile popover
```

### CODE REVIEW

File: plugins/call_button/call_button.tsx
Comment (streamer45) [inline plugins/call_button/call_button.tsx:37]: This seems a touch overly complex. I'd expect to pass a channel ID where the call should be started in and not much more as anything else should be likely dealt with by the caller.

Maybe we could even re-purpose the `currentChannel` arg for this without breaking anything. What do you think?

**Review by streamer45 — COMMENTED** [review]
Nicely done, works great :tada: I only left one possible suggestion.

As an aside, it would be great to write an e2e test to cover this. If this is something you'd be interested in contributing just let me know :)

File: plugins/call_button/call_button.tsx
Comment (sinansonmez) [reply plugins/call_button/call_button.tsx:37]: @streamer45  This is exactly what I was thinking initially. I thought I just need to pass channel to start a call. Then I realized it wasnt that easy. 

IMO re-purposing `currentChannel` is possible where current user has already a DM channel with the other user. However, this is not always the case. And if there is no DM channel exists, I need to create DM Channel only if user clicks the call button, not on component mount. If I create DM channel in `index` file, an unnecessary new DM channel is created everytime when user opens profile popover

That is why implemented logic like this:
- pass `channelToStartCall` i.e. already existing DM Channel
- if `channelToStartCall` doesn't exist, create a DM Channel using `currentUserId` and `userId` which are passed as prop on user click call button

**Comment by sinansonmez** [comment]
@streamer45 sure, I can write e2e tests to cover the case. If you open another ticket for e2e test, just assign to me, or I can add the tests to this PR.

File: plugins/call_button/call_button.tsx
Comment (streamer45) [reply plugins/call_button/call_button.tsx:37]: Thanks for the added context @sinansonmez . 

I feel most of the complication is coming from the fact that we are slightly abusing the plugin button. Maybe the simplest solution would be to stop relying on that but implement the logic directly in the profile popover. 

After all you could always access the callback needed to join a call from `state.plugins.components.CallButton[0].action` (with the appropriate undefined checks of course) and use that in a click handler of your custom button.

I understand this proposal would mean reverting most of the work done so far but it honestly feels cleaner to me. That said, happy to wait to hear @AshishDhama's thoughts given his superior experience on this side of the app.

Comment (sinansonmez) [reply plugins/call_button/call_button.tsx:37]: @streamer45 That is an interesting (also ref. Zef's Platformer 😄 ) to point of view. 

No worries about reverting the changes. For me the important thing is to impletement it in a best possible way 👨‍💻 

I would wait for Ashish's feedback and try to make it more cleaner

**Comment by sinansonmez** [comment]
@streamer45 e2e test is added. I just need a test case id which @jgilliam17 might help.

**Comment by sinansonmez** [comment]
@jgilliam17 I see that the test is failing because of `Calls cant be initiated in an unsecure`

File: plugins/call_button/call_button.tsx
Comment (AshishDhama) [reply plugins/call_button/call_button.tsx:37]: @streamer45 I agree we should move this logic out of the `call_button.tsx`, but in my opinion, profile popover isn’t the right place either. I would vote for a custom button or hook handling this logic instead of putting it in the `call_button.tsx` or `profile_popover.tsx`. cc// @sinansonmez 
@streamer45 From the logic in `call_button.tsx.`, we expect multiple call plugins, right? If we are expecting a multiple-call plugin I think we will be rewriting the button UI, In that case, I think some modifications in `call_button.tsx` and a custom hook would be best.
But if in the `profile popover` we just want our calls plugin button I would say let’s create a custom button component and use that. @laneycs could you please guide us here?

Comment (AshishDhama) [reply plugins/call_button/call_button.tsx:37]: TL;DR Move the logic to create DM, out of `call_button.tsx` to either hook or custom button.

File: e2e/cypress/tests/integration/enterprise/profile_popover/profile_popover_spec.js
Comment (jgilliam17) [inline e2e/cypress/tests/integration/enterprise/profile_popover/profile_popover_spec.js:427]:
```suggestion
    it('MM-T5382 Call User - Call should start in DM channel', () => {
```
adding test ID

**Comment by jgilliam17** [comment]
> @jgilliam17 I see that the test is failing because of `Calls cant be initiated in an unsecure` (see the picture in the link below). Is there anything that I can enable/ disable in the setting to bypass this check, maybe @streamer45 can help
> 
> https://automation-dashboard.vercel.app/_next/image?url=https%3A%2F%2Fmm-automation-s3uploadbucket-1c95bzy9fbgeq.s3.amazonaws.com%2Fmattermost-webapp%2Fwebapp-pr-12080%2F295665-d2bc590-onprem-ent%2F0b88a206-e601-4730-ae05-5451911e8454.png&w=1920&q=75

@saturninoabril Do you have a suggestion on how to unblock the test? Thanks

**Comment by streamer45** [comment]
Calls generally require HTTPs to work unless it's pointing to `localhost`. Maybe the simplest thing to do would be to move/convert the e2e test to the Calls suite (https://github.com/mattermost/mattermost-plugin-calls/tree/main/e2e).

File: plugins/call_button/call_button.tsx
Comment (sinansonmez) [reply plugins/call_button/call_button.tsx:37]: @laneycs please let me know what do you think about 2 potential options mentioned by Ashish

Comment (streamer45) [reply plugins/call_button/call_button.tsx:37]: > From the logic in call_button.tsx., we expect multiple call plugins, right?

Correct, the button is meant to support multiple plugins (e.g. Zoom, Jitsi) with Calls being the default one. That said I really don't know if the same is expected from the popover. My guess would be no as we also trying to integrate Calls more into the main product and at some point it will likely be part of the suite.

To unblock this, since it's a bug fix, I propose to extract some of the logic as @AshishDhama suggested and create a custom button component. @sinansonmez Let me know if that makes sense.

Comment (sinansonmez) [reply plugins/call_button/call_button.tsx]: Thanks @streamer45 this makes sense. I reverted the previous changes in `call_button.tsx` and I created a custom component for `ProfilePopover`

**Review by streamer45 — APPROVED** [review]
Looks great @sinansonmez :100: Given it's a bit tricky to test Calls from here, could you revert the e2e portion to avoid failures? We can easily add an e2e test for this functionality on the Calls side once this is merged.

**Comment by sinansonmez** [comment]
thanks @streamer45 . I reverted e2e tests as well

**Comment by streamer45** [comment]
@AshishDhama @itao Gentle ping on this, thanks!

File: components/profile_popover_call_button/profile_popover_call_button.tsx
Comment (AshishDhama) [inline components/profile_popover_call_button/profile_popover_call_button.tsx:64]: I think we can avoid undefined by doing something like this
```suggestion
            onClick={clickHandler}
```

```typescript
const handleStartCall = async () => {
        const channelForCall = await getDmChannel();
        item.action?.(channelForCall, channelMember);
    };
    const clickHandler = () => { 
    if(clickEnabled){
    handleStartCall
 }
}
```

**Review by AshishDhama — CHANGES_REQUESTED** [review]
Most of it looks good to me except one change

**Review by itao — APPROVED** [review]

**Comment by sinansonmez** [comment]
thanks @furqanmlk for taking a loot at the PR. For now, test server will fail to create and e2e tests wont start due to failing `ci/upload-s3` step which cause by recent migration to github actions. As mentioned by Saturnino, Release team is working on the solution: https://github.com/mattermost/mattermost-webapp/pull/11262#issuecomment-1433889621

File: components/profile_popover_call_button/profile_popover_call_button.tsx
Comment (sinansonmez) [reply components/profile_popover_call_button/profile_popover_call_button.tsx]: Thanks for the suggestion @AshishDhama. updated as suggested

**Comment by furqanmlk** [comment]
> thanks @furqanmlk for taking a loot at the PR. For now, test server will fail to create and e2e tests wont start due to failing `ci/upload-s3` step which cause by recent migration to github actions. As mentioned by Saturnino, Release team is working on the solution: [#11262 (comment)](https://github.com/mattermost/mattermost-webapp/pull/11262#issuecomment-1433889621)

Thanks @sinansonmez, for confirming I will wait until e2e pipeline is fixed.

**Review by AshishDhama — APPROVED** [review]

**Comment by furqanmlk** [comment]
@sinansonmez I can run e2e by creating a draft PR out of this PR.
However, can you please fix the conflict?

**Comment by sinansonmez** [comment]
thanks @furqanmlk merge conflicts are fixed

**Comment by sinansonmez** [comment]
@streamer45 please merge if everything is OK

**Comment by amyblais** [comment]
@streamer45 Please help open a PR to cherry-pick to ``release-7.9``.

**Pull request merged**
